File: api/invoices.php
<?php
header('Content-Type: application/json');
header('Access-Control-Allow-Origin: *');
header('Access-Control-Allow-Methods: POST, GET, PUT, DELETE, OPTIONS');
header('Access-Control-Allow-Headers: Content-Type');

if ($_SERVER['REQUEST_METHOD'] === 'OPTIONS') exit(0);

// Start session before including auth
if (session_status() === PHP_SESSION_NONE) {
    session_start();
}

require_once __DIR__ . '/../config/database.php';

$conn = getConnection();
$action = $_GET['action'] ?? '';

// Simple auth check function
function checkAuth() {
    return isset($_SESSION['logged_in']) && $_SESSION['logged_in'] === true;
}

function getCurrentUserId() {
    return $_SESSION['user_id'] ?? 1;
}

switch ($action) {
    case 'list':
        $status = isset($_GET['status']) ? $conn->real_escape_string($_GET['status']) : '';
        $where = $status ? "WHERE i.payment_status = '$status'" : "";
        
        $sql = "SELECT i.*, c.name as customer_name FROM invoices i 
                LEFT JOIN customers c ON i.customer_id = c.id $where
                ORDER BY i.created_at DESC LIMIT 100";
        $result = $conn->query($sql);
        $invoices = [];
        if ($result) {
            while ($row = $result->fetch_assoc()) $invoices[] = $row;
        }
        echo json_encode(['success' => true, 'data' => $invoices]);
        break;
        
    case 'get':
        $id = intval($_GET['id'] ?? 0);
        $sql = "SELECT i.*, c.name as customer_name, c.phone as customer_phone, c.address as customer_address
                FROM invoices i LEFT JOIN customers c ON i.customer_id = c.id WHERE i.id = $id";
        $result = $conn->query($sql);
        if ($result && $result->num_rows === 1) {
            $invoice = $result->fetch_assoc();
            $itemsResult = $conn->query("SELECT ii.*, p.name as product_name FROM invoice_items ii 
                                         LEFT JOIN products p ON ii.product_id = p.id 
                                         WHERE ii.invoice_id = $id");
            $items = [];
            if ($itemsResult) {
                while ($row = $itemsResult->fetch_assoc()) $items[] = $row;
            }
            $invoice['items'] = $items;
            echo json_encode(['success' => true, 'data' => $invoice]);
        } else {
            echo json_encode(['success' => false, 'message' => 'Invoice not found']);
        }
        break;
        
    case 'create':
        if (!checkAuth()) {
            echo json_encode(['success' => false, 'message' => 'Please login first']);
            break;
        }
        
        $data = json_decode(file_get_contents('php://input'), true);
        
        if (!$data) {
            echo json_encode(['success' => false, 'message' => 'Invalid data']);
            break;
        }
        
        $customerId = !empty($data['customer_id']) ? intval($data['customer_id']) : 'NULL';
        $items = $data['items'] ?? [];
        $taxRate = floatval($data['tax_rate'] ?? 0);
        $discountAmount = floatval($data['discount_amount'] ?? 0);
        $paymentMethod = $conn->real_escape_string($data['payment_method'] ?? 'cash');
        $paymentStatus = $conn->real_escape_string($data['payment_status'] ?? 'paid');
        
        if (empty($items)) { 
            echo json_encode(['success' => false, 'message' => 'No items in cart']); 
            break; 
        }
        
        // Calculate totals
        $subtotal = 0;
        foreach ($items as $item) {
            $subtotal += floatval($item['quantity']) * floatval($item['unit_price']);
        }
        $taxAmount = $subtotal * ($taxRate / 100);
        $totalAmount = $subtotal + $taxAmount - $discountAmount;
        
        // Generate invoice number
        $prefix = 'INV';
        $dateStr = date('Ymd');
        $countResult = $conn->query("SELECT COUNT(*) as cnt FROM invoices WHERE DATE(created_at) = CURDATE()");
        $count = 1;
        if ($countResult) {
            $row = $countResult->fetch_assoc();
            $count = intval($row['cnt']) + 1;
        }
        $invoiceNumber = $prefix . $dateStr . str_pad($count, 4, '0', STR_PAD_LEFT);
        
        $userId = getCurrentUserId();
        
        // Insert invoice
        $conn->begin_transaction();
        try {
            $sql = "INSERT INTO invoices (invoice_number, customer_id, user_id, subtotal, tax_rate, tax_amount, 
                    discount_amount, total_amount, payment_method, payment_status) 
                    VALUES ('$invoiceNumber', $customerId, $userId, $subtotal, $taxRate, $taxAmount, 
                    $discountAmount, $totalAmount, '$paymentMethod', '$paymentStatus')";
            
            if (!$conn->query($sql)) {
                throw new Exception('Failed to create invoice: ' . $conn->error);
            }
            
            $invoiceId = $conn->insert_id;
            
            // Insert items and update stock
            foreach ($items as $item) {
                $productId = intval($item['product_id']);
                $quantity = intval($item['quantity']);
                $unitPrice = floatval($item['unit_price']);
                $totalPrice = $quantity * $unitPrice;
                
                // Get product name
                $prodResult = $conn->query("SELECT name FROM products WHERE id = $productId");
                $productName = 'Unknown';
                if ($prodResult && $prodResult->num_rows > 0) {
                    $prod = $prodResult->fetch_assoc();
                    $productName = $conn->real_escape_string($prod['name']);
                }
                
                $itemSql = "INSERT INTO invoice_items (invoice_id, product_id, product_name, quantity, unit_price, total_price)
                            VALUES ($invoiceId, $productId, '$productName', $quantity, $unitPrice, $totalPrice)";
                
                if (!$conn->query($itemSql)) {
                    throw new Exception('Failed to add item: ' . $conn->error);
                }
                
                // Update stock
                $conn->query("UPDATE products SET stock_quantity = GREATEST(0, stock_quantity - $quantity) WHERE id = $productId");
            }
            
            $conn->commit();
            echo json_encode(['success' => true, 'invoice_id' => $invoiceId, 'invoice_number' => $invoiceNumber]);
            
        } catch (Exception $e) {
            $conn->rollback();
            echo json_encode(['success' => false, 'message' => $e->getMessage()]);
        }
        break;
        
    case 'update_status':
        if (!checkAuth()) {
            echo json_encode(['success' => false, 'message' => 'Unauthorized']);
            break;
        }
        $data = json_decode(file_get_contents('php://input'), true);
        $id = intval($data['id'] ?? 0);
        $status = $conn->real_escape_string($data['status'] ?? '');
        $conn->query("UPDATE invoices SET payment_status = '$status' WHERE id = $id");
        echo json_encode(['success' => true]);
        break;
        
    case 'today_summary':
        $sql = "SELECT COUNT(*) as total_invoices,
                COALESCE(SUM(CASE WHEN payment_status = 'paid' THEN total_amount ELSE 0 END), 0) as paid_amount,
                COALESCE(SUM(total_amount), 0) as total_amount
                FROM invoices WHERE DATE(created_at) = CURDATE()";
        $result = $conn->query($sql);
        $data = $result ? $result->fetch_assoc() : ['total_invoices' => 0, 'paid_amount' => 0, 'total_amount' => 0];
        echo json_encode(['success' => true, 'data' => $data]);
        break;
        
    default:
        echo json_encode(['success' => false, 'message' => 'Invalid action']);
}

$conn->close();
?>

File: routes/invoices.py
"""
Invoices API Routes
BillMaster Pro - Python/Flask Backend (SQLite)
"""

from flask import Blueprint, request, jsonify, session
from datetime import datetime
import sys
import os

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from config.database import get_connection, dict_from_row, dict_list_from_rows, generate_invoice_number

invoices_bp = Blueprint('invoices', __name__)

def is_logged_in():
    return session.get('logged_in', False)

def get_current_user_id():
    return session.get('user_id', 1)

@invoices_bp.route('/invoices.php', methods=['GET', 'POST', 'PUT', 'DELETE', 'OPTIONS'])
def invoices_handler():
    """Handle invoice requests - maintains PHP-style URL for frontend compatibility"""
    if request.method == 'OPTIONS':
        return '', 200
    
    action = request.args.get('action', '')
    
    if action == 'list':
        return list_invoices()
    elif action == 'get':
        return get_invoice()
    elif action == 'create':
        return create_invoice()
    elif action == 'update_status':
        return update_status()
    elif action == 'today_summary':
        return today_summary()
    else:
        return jsonify({'success': False, 'message': 'Invalid action'})

def list_invoices():
    """List all invoices with optional status filter"""
    try:
        status = request.args.get('status', '')
        
        conn = get_connection()
        if not conn:
            return jsonify({'success': False, 'message': 'Database connection failed'})
        
        cursor = conn.cursor()
        
        if status:
            cursor.execute("""
                SELECT i.*, c.name as customer_name 
                FROM invoices i 
                LEFT JOIN customers c ON i.customer_id = c.id
                WHERE i.payment_status = ?
                ORDER BY i.created_at DESC 
                LIMIT 100
            """, (status,))
        else:
            cursor.execute("""
                SELECT i.*, c.name as customer_name 
                FROM invoices i 
                LEFT JOIN customers c ON i.customer_id = c.id
                ORDER BY i.created_at DESC 
                LIMIT 100
            """)
        
        invoices = dict_list_from_rows(cursor.fetchall())
        cursor.close()
        conn.close()
        
        return jsonify({'success': True, 'data': invoices})
        
    except Exception as e:
        return jsonify({'success': False, 'message': str(e)})

def get_invoice():
    """Get single invoice with items"""
    try:
        inv_id = request.args.get('id', 0, type=int)
        
        conn = get_connection()
        if not conn:
            return jsonify({'success': False, 'message': 'Database connection failed'})
        
        cursor = conn.cursor()
        
        cursor.execute("""
            SELECT i.*, c.name as customer_name, c.phone as customer_phone, c.address as customer_address
            FROM invoices i 
            LEFT JOIN customers c ON i.customer_id = c.id 
            WHERE i.id = ?
        """, (inv_id,))
        row = cursor.fetchone()
        invoice = dict_from_row(row) if row else None
        
        if invoice:
            # Get invoice items
            cursor.execute("""
                SELECT ii.*, p.name as product_name 
                FROM invoice_items ii 
                LEFT JOIN products p ON ii.product_id = p.id 
                WHERE ii.invoice_id = ?
            """, (inv_id,))
            items = dict_list_from_rows(cursor.fetchall())
            invoice['items'] = items
            
            cursor.close()
            conn.close()
            return jsonify({'success': True, 'data': invoice})
        else:
            cursor.close()
            conn.close()
            return jsonify({'success': False, 'message': 'Invoice not found'})
            
    except Exception as e:
        return jsonify({'success': False, 'message': str(e)})

def create_invoice():
    """Create new invoice with items"""
    if not is_logged_in():
        return jsonify({'success': False, 'message': 'Please login first'})
    
    try:
        data = request.get_json()
        
        if not data:
            return jsonify({'success': False, 'message': 'Invalid data'})
        
        customer_id = data.get('customer_id')
        items = data.get('items', [])
        tax_rate = float(data.get('tax_rate', 0))
        discount_amount = float(data.get('discount_amount', 0))
        payment_method = data.get('payment_method', 'cash')
        payment_status = data.get('payment_status', 'paid')
        
        if not items:
            return jsonify({'success': False, 'message': 'No items in cart'})
        
        # Calculate totals
        subtotal = sum(float(item['quantity']) * float(item['unit_price']) for item in items)
        tax_amount = subtotal * (tax_rate / 100)
        total_amount = subtotal + tax_amount - discount_amount
        
        conn = get_connection()
        if not conn:
            return jsonify({'success': False, 'message': 'Database connection failed'})
        
        cursor = conn.cursor()
        
        # Generate invoice number
        invoice_number = generate_invoice_number(conn)
        user_id = get_current_user_id()
        
        try:
            # Insert invoice
            cursor.execute("""
                INSERT INTO invoices (invoice_number, customer_id, user_id, subtotal, tax_rate, tax_amount, 
                                     discount_amount, total_amount, payment_method, payment_status) 
                VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
            """, (invoice_number, customer_id, user_id, subtotal, tax_rate, tax_amount,
                  discount_amount, total_amount, payment_method, payment_status))
            
            invoice_id = cursor.lastrowid
            
            # Insert items and update stock
            for item in items:
                product_id = int(item['product_id'])
                quantity = int(item['quantity'])
                unit_price = float(item['unit_price'])
                total_price = quantity * unit_price
                
                # Get product name
                cursor.execute("SELECT name FROM products WHERE id = ?", (product_id,))
                prod = cursor.fetchone()
                product_name = prod['name'] if prod else 'Unknown'
                
                # Insert invoice item
                cursor.execute("""
                    INSERT INTO invoice_items (invoice_id, product_id, product_name, quantity, unit_price, total_price)
                    VALUES (?, ?, ?, ?, ?, ?)
                """, (invoice_id, product_id, product_name, quantity, unit_price, total_price))
                
                # Update stock
                cursor.execute("""
                    UPDATE products SET stock_quantity = MAX(0, stock_quantity - ?) WHERE id = ?
                """, (quantity, product_id))
            
            conn.commit()
            cursor.close()
            conn.close()
            
            return jsonify({
                'success': True, 
                'invoice_id': invoice_id, 
                'invoice_number': invoice_number
            })
            
        except Exception as e:
            conn.rollback()
            cursor.close()
            conn.close()
            return jsonify({'success': False, 'message': str(e)})
        
    except Exception as e:
        return jsonify({'success': False, 'message': str(e)})

def update_status():
    """Update invoice payment status"""
    if not is_logged_in():
        return jsonify({'success': False, 'message': 'Unauthorized'})
    
    try:
        data = request.get_json() or {}
        inv_id = data.get('id', 0)
        status = data.get('status', '')
        
        conn = get_connection()
        if not conn:
            return jsonify({'success': False, 'message': 'Database connection failed'})
        
        cursor = conn.cursor()
        cursor.execute("UPDATE invoices SET payment_status = ? WHERE id = ?", (status, inv_id))
        conn.commit()
        cursor.close()
        conn.close()
        
        return jsonify({'success': True})
        
    except Exception as e:
        return jsonify({'success': False, 'message': str(e)})

def today_summary():
    """Get today's invoice summary"""
    try:
        conn = get_connection()
        if not conn:
            return jsonify({'success': False, 'message': 'Database connection failed'})
        
        cursor = conn.cursor()
        cursor.execute("""
            SELECT 
                COUNT(*) as total_invoices,
                COALESCE(SUM(CASE WHEN payment_status = 'paid' THEN total_amount ELSE 0 END), 0) as paid_amount,
                COALESCE(SUM(total_amount), 0) as total_amount
            FROM invoices 
            WHERE DATE(created_at) = DATE('now')
        """)
        row = cursor.fetchone()
        data = dict_from_row(row) if row else {'total_invoices': 0, 'paid_amount': 0, 'total_amount': 0}
        cursor.close()
        conn.close()
        
        return jsonify({'success': True, 'data': data})
        
    except Exception as e:
        return jsonify({'success': False, 'message': str(e)})
